code/modules/projectiles/guns/energy.dm

```diff
@@ -29,69 +30,150 @@
 				return 0								//Note, uses /10 because of shitty mods to the cell system
+		power_supply.give(E.e_cost) //... to recharge the shot
+		update_icon()
 
-		power_supply.give(charge_cost) //... to recharge the shot
+/obj/item/weapon/gun/energy/attack_self(mob/living/user as mob)
+	if(ammo_type.len > 1)
+		select_fire(user)
 		update_icon()
-	return 1
+
+/obj/item/weapon/gun/energy/afterattack(atom/target as mob|obj|turf, mob/living/user as mob|obj, params)
+	newshot() //prepare a new shot
+	..()
+
+/obj/item/weapon/gun/energy/can_shoot()
+	var/obj/item/ammo_casing/energy/shot = ammo_type[select]
+	return power_supply.charge >= shot.e_cost
+
+/obj/item/weapon/gun/energy/newshot()
+	if (!ammo_type || !power_supply)
+		return
+	var/obj/item/ammo_casing/energy/shot = ammo_type[select]
+	if(power_supply.charge >= shot.e_cost) //if there's enough power in the power_supply cell...
+		chambered = shot //...prepare a new shot based on the current ammo type selected
+		chambered.newshot()
+	return
+
+/obj/item/weapon/gun/energy/process_chamber()
+	if(chambered && !chambered.BB) //if BB is null, i.e the shot has been fired...
+		var/obj/item/ammo_casing/energy/shot = chambered
+		power_supply.use(shot.e_cost)//... drain the power_supply cell
+	chambered = null //either way, released the prepared shot
+	return
+
+/obj/item/weapon/gun/energy/proc/select_fire(mob/living/user)
+	select++
+	if (select > ammo_type.len)
+		select = 1
+	var/obj/item/ammo_casing/energy/shot = ammo_type[select]
+	fire_sound = shot.fire_sound
+	fire_delay = shot.delay
+	if (shot.select_name)
+		to_chat(user, "<span class='notice'>[src] is now set to [shot.select_name].</span>")
+	update_icon()
+	return
+
+/obj/item/weapon/gun/energy/update_icon()
+	overlays.Cut()
+	var/ratio = Ceiling((power_supply.charge / power_supply.maxcharge) * charge_sections)
+	var/obj/item/ammo_casing/energy/shot = ammo_type[select]
+	var/iconState = "[icon_state]_charge"
+	var/itemState = null
+	if(!initial(item_state))
+		itemState = icon_state
+	if (modifystate)
+		overlays += "[icon_state]_[shot.select_name]"
+		iconState += "_[shot.select_name]"
+		if(itemState)
+			itemState += "[shot.select_name]"
+	if(power_supply.charge < shot.e_cost)
+		overlays += "[icon_state]_empty"
+	else
+		if(!shaded_charge)
+			for(var/i = ratio, i >= 1, i--)
+				overlays += image(icon = icon, icon_state = icon_state, pixel_x = ammo_x_offset * (i -1))
+		else
+			if(modifystate)
+				overlays += image(icon = icon, icon_state = "[icon_state]_[shot.select_name]_charge[ratio]")
+			else
+				overlays += image(icon = icon, icon_state = "[icon_state]_charge[ratio]")
+	if(F && can_flashlight)
+		var/iconF = "flight"
+		if(F.on)
+			iconF = "flight_on"
+		overlays += image(icon = icon, icon_state = iconF, pixel_x = flight_x_offset, pixel_y = flight_y_offset)
+	if(itemState)
+		itemState += "[ratio]"
+		item_state = itemState
+
+/obj/item/weapon/gun/energy/ui_action_click()
+	toggle_gunlight()

```

Comment:
Earlier, I didn't see where the action button for guns got added----did I miss it, or is this an oversight/it wasn't added?